Document less known EV3 color sensor modes

JakubVanek opened the pull request:
Fixes https://github.com/ev3dev/ev3dev/issues/1386
References:
* REF-RAW: https://github.com/simonedegiacomi/EV3-API/commit/20a57f7b944a1e53477ebdfe61f96ebf2d6c66b7#r37442226
* RGB-RAW: https://github.com/simonedegiacomi/EV3-API/commit/8562bd29da2ab430807741f7374ac60434eb793e#r37442413
* COL-CAL: https://github.com/ev3dev/ev3dev/issues/1386

JakubVanek committed:
sensors: document the EV3 color's REF-RAW mode

More info: https://github.com/simonedegiacomi/EV3-API/commit/20a57f7b944a1e53477ebdfe61f96ebf2d6c66b7#r37442226

Signed-off-by: Jakub Vanek <linuxtardis@gmail.com>

JakubVanek committed:
sensors: document the EV3 color's RGB-RAW mode

More info: https://github.com/simonedegiacomi/EV3-API/commit/8562bd29da2ab430807741f7374ac60434eb793e#r37442413

Signed-off-by: Jakub Vanek <linuxtardis@gmail.com>

JakubVanek committed:
sensors: document the EV3 color's COL-CAL mode

More info: https://github.com/ev3dev/ev3dev/issues/1386

Signed-off-by: Jakub Vanek <linuxtardis@gmail.com>

dlech commented on sensors/ev3_uart_sensor_defs.c:
```diff
@@ -86,25 +86,67 @@ const struct ev3_uart_sensor_info ev3_uart_sensor_defs[] = {
+				 *    illumination will cause lower ADC
+				 *    measurements., so to obtain reflected
+				 *    light intensity similar to
+				 *    ``COL-REFLECT``, it is necessary to
```
This is markdown, so single backtick should be sufficient.

JakubVanek replied:
I have found double backticks in the surrounding documentation (for example around line 313):
```
 * .. [#lego-ev3-gyro-mode0-value0-overflow] If
 *    you spin around too many times in ``GYRO-ANG``
 *    or ``GYRO-G&A`` mode, it will get stuck at
 *    32767 or overflow through -32768 depending
 *    on when the sensor was manufactured.
```

dlech commented on sensors/ev3_uart_sensor_defs.c:
```diff
@@ -86,25 +86,67 @@ const struct ev3_uart_sensor_info ev3_uart_sensor_defs[] = {
 				.name		= "REF-RAW",
 				.data_sets	= 2,
 				.data_type	= LEGO_SENSOR_DATA_S16,
+				.units		= "ADC",
```
This should exactly match what the sensor sends over the UART during initialization. My notes show this mode does not have units. Have you seen different?

JakubVanek replied:
No, I haven't seen this, I thought this is only for documentation purposes. I will remove it.

dlech commented on sensors/ev3_uart_sensor_defs.c:
```diff
@@ -86,25 +86,67 @@ const struct ev3_uart_sensor_info ev3_uart_sensor_defs[] = {
+				 *    but they are not precalibrated.
+				 *
+				 * .. [#lego-ev3-color-mode4-value3] This value
+				 *    is similar to the REF-RAW ambient
```
Backticks around `REF-RAW` to be consistent?

JakubVanek replied:
Yes, I will add them.

dlech commented on sensors/ev3_uart_sensor_defs.c:
```diff
@@ -86,25 +86,67 @@ const struct ev3_uart_sensor_info ev3_uart_sensor_defs[] = {
+				 * @value1_footnote: [#lego-ev3-color-mode4-value012]_
+				 * @value2: Reflected blue light intensity (0 to 1023)
+				 * @value2_footnote: [#lego-ev3-color-mode4-value012]_
+				 * @value3: ADC reading for LEDs off (reference) (0 to 1023)
```
I've been wondering if I should revert https://github.com/ev3dev/lego-linux-drivers/commit/3f7760ec7be41813a55dfd911a529606b38e1538. The sensor itself says that this mode only has 3 values. I got the idea about the 4th value from Dexter Industries BrickPi python library. But I've been wondering it this is actually a useful value or just leftover data in a buffer. UART sensors can only send data in sizes that are a power of 2, so that could explain why the extra data is sent even if it was never meant to be used.

Have you actually tried the ev3dev-buster kernel to see if this value seems similar to the same value in the RAW-REF mode?

JakubVanek replied:
I haven't tried this on ev3dev, but on stock firmware with c4ev3 it seemed to work similarly. I will recheck it. There is a simple test for this - if this value briefly changes when the intensity of external lighting illuminating the sensor changes, then this should be the ADC measurement when no LEDs are on.

I think that this value could still be useful - while one cannot measure steady ambient light, one could measure changes in ambient light this way.

dlech replied:
OK, if it seems useful, then let's keep it. :smile:

JakubVanek replied:
I have tried it again on C4EV3 (I don't have a SD card by me... :smile:) and the fourth value indeed behaves like the ambient of REF-RAW. I think this value can be useful and it's nice that ev3dev would provide it out of the box.

Without changes in external lighting, the values move around 670. When I move my phone's flashlight in front of the sensor, it temporarily moves to zero and the returns back. When I remove the light, the value jumps up and then returns again.

dlech commented on sensors/ev3_uart_sensor_defs.c:
```diff
@@ -86,25 +86,67 @@ const struct ev3_uart_sensor_info ev3_uart_sensor_defs[] = {
 				.name		= "RGB-RAW",
 				.data_sets	= 4,
 				.data_type	= LEGO_SENSOR_DATA_S16,
+				.units		= "ADC",
```
Same here, this mode does not have units.

JakubVanek replied:
Yup, I will remove it.

dlech commented on sensors/ev3_uart_sensor_defs.c:
```diff
@@ -113,13 +155,19 @@ const struct ev3_uart_sensor_info ev3_uart_sensor_defs[] = {
-				 * @value1: ???
-				 * @value2: ???
-				 * @value3: ???
+				 * @value0: ??? (0 to 1023???)
```
According to my notes, this mode has 65535 reported as the max value (both raw and SI units).

JakubVanek replied:
Yes, that is consistent with LEGO's `typedata.rcf`, I was guessing that this is coming from ADC, so >1023 wouldn't make sense. I will remove the change.

JakubVanek replied:
On the stock firmware the values indeed exceed 1023, so the 65535 upper bound is likely to be true.

JakubVanek replied:
Added 65535 as the upper bound :+1:

dlech reviewed (commented):
Thanks for the PR. I made a few comments/questions to have a look at.

dlech commented on sensors/ev3_uart_sensor_defs.c:
```diff
@@ -86,36 +86,87 @@ const struct ev3_uart_sensor_info ev3_uart_sensor_defs[] = {
 				.name		= "REF-RAW",
 				.data_sets	= 2,
 				.data_type	= LEGO_SENSOR_DATA_S16,
+				.units		= "",
```
I don't think we need to add the units field here. This file already has the convention of using the default value for omitted fields.

JakubVanek replied:
Fixed

JakubVanek committed:
sensors: hide last value of the COL-CAL mode

It seems that this value is always zero - it
is not even shown in the test calibration application
on the stock firmware.

Signed-off-by: Jakub Vanek <linuxtardis@gmail.com>

JakubVanek committed:
sensors: use U16 data format for the COL-CAL mode

Experimentation suggests that the values are only 16bit integers.
The maximum of 65535 futher hints towards that this is a uint16_t.

Signed-off-by: Jakub Vanek <linuxtardis@gmail.com>

dlech commented:
merged, thanks!
